site/en/docs/extensions/mv3/promises/index.md

```diff
@@ -0,0 +1,214 @@
+* Where error handling would be too difficult using callbacks.
+* When you want a simpler way to invoke a number of concurrent methods and gather the results into a single thread of code.
+
+### Converting a callback to a promise {: #compare-to-callback}
+
+One way to understand how you can use promises in extensions APIs is to compare two equivalent code
+fragments, one using a callback and one using a promise. The following example shows this
+comparison:
+
+<!--
+// --- Standard callback implementation ---
+// --- Promise implementation ---
+-->
+
+#### Standard callback implementation
+
+```js
+
+function openTabOnRight(onComplete) {
+  chrome.tabs.query(queryOptions, function(tabs) {
+    if (chrome.runtime.lastError) {
+      onComplete({error: chrome.runtime.lastError});
+      return;
+    }
+
+    if (!tabs.length) {
+      onComplete();
+      return;
+    };
+
+    chrome.tabs.create({
+      url: 'https://example.com',
+      index: tab[0].index + 1,
+    }, function(tab) {
+      if (chrome.runtime.lastError) {
+        onComplete({error: chrome.runtime.lastError});
+      }
+
+      console.log('tab created', tab);
+      onComplete(tab);
+    });
+  });
+}
+```
+
+#### Promise implementation
+
+```js
+function openTabOnRight() {
+  // Errors are automatically propagated down the promise chain
+  return chrome.tabs.query(queryOptions)
+    .then((tabs) => {
+      if (!tabs.length) return;
+
+      return chrome.tabs.create({
+        url: 'https://example.com',
+        index: tab[0].index + 1,
+      });
+    })
+    .then(tab => {
+      if (!tab) return;
+
+      console.log('tab created', tab);
+      return tab;
+    });
+}
+```
+
+
+<!-- notes here about the code fragments -->
+
+### Error handling
+
+Returning errors works differently depending on if the extension is using a callback or a promise.
+
+#### Error handling with Callbacks
+
+If using a callback, then `chrome.runtime.lastError` is set for the duration of the execution of the
+callback. It is not thrown as a JS Error (which would interrupt JS execution), and is not set
+outside the duration of the callback run (which would result in it being "randomly" set during other
+execution).  The extension would look at the last error like this:
+
+```js
+chrome.tabs.create({...}, (result) => {
+  if (chrome.runtime.lastError) {
+    // Handle last error
+  }
+});
+```
+
+Promises are designed to deliver asynchronous results, both success and failure.  A failure in a
+promise (a promise rejection) is handled differently.  It might look like this:
+
+```js
+chrome.tabs.create({...})
+  .then((result) => {
+    // success case
+  })
+  .catch((e) => {
+    // failure case
+  });
+```
+
+Instead of setting `chrome.runtime.lastError`, we deliver the error directly to the "catch" of the
+promise. We don't set the `lastError` in the success case, because the onFulfilled handler (the
+argument in the `.then())` is never invoked in the failure case.
```

Comment:
Pinged Tim for clarification on how Chrome handles `lastError` in promises, as the current copy left a bit of a gap for me. 

The second sentence seems a little confused. Is `lastError` referring to `chorme.runtime.lastError` or the argument passed into the `.catch()` handler? It's also not clear to me why the onFufilled handler (this is the first time that name has appeared, might be better to say "success handler"?) is mentioned.

EDIT: I tend to read pretty linearly and I only just say the below aside. I think this paragraph can be simplified.

Reply:
Pruned this for simplicity in 9e0faef.